elki-core-math/src/main/java/elki/math/statistics/dependence/mcde/MWPTest.java

```diff
@@ -139,53 +142,81 @@ private MWPTest() {
    * Framework.
    *
    * @param start Starting index value for statistical test
-   * @param end End index value for statistical test
+   * @param width Width of the slice for statistical test (endindex = start + width)
    * @param slice Return value of randomSlice() created with the index that is
    *        not for the reference dimension
    * @param corrected_ranks Index of the reference dimension, return value of
    *        corrected_ranks() computed for reference dimension
    * @return p-value from two sided Mann-Whitney-U test
    */
-  public double statisticalTest(int start, int end, boolean[] slice, MWPRanking corrected_ranks) {
+  public double statisticalTest(int start, int width, boolean[] slice, MWPRanking corrected_ranks) {
+    final int safeStart = getSafeCut(start, corrected_ranks);
+    final int len = corrected_ranks.len;
+    final int sliceEndSearchStart = safeStart + width > len -1 ? len -1 : safeStart + width;
+    final int safeEnd = getSafeCut(sliceEndSearchStart, corrected_ranks);
+
     double R = 0.0;
     int n1 = 0;
-    for(int j = start; j < end; j++) {
+    for(int j = safeStart; j < safeEnd; j++) {
       if(slice[corrected_ranks.index[j]]) {
         R += corrected_ranks.adjusted[j];
         n1++;
       }
     }
 
-    // Cancel the offset in case the marginal restriction does not start from 0
-    // see "acc-(cutStart*count)" in reference implementation of MWP
-    R -= start * n1;
+    // This is to cancel the offset in case the marginal restriction does not
+    // start from 0 see "acc-(cutStart*count)" is reference implementation of
+    // MWP
+    R -= safeStart * n1;
 
-    final int cutLength = end - start;
+    final int cutLength = safeEnd - safeStart;
     if(n1 == 0 || n1 == cutLength) {
       return 1;
     }
 
     final double U = R - 0.5 * n1 * (n1 - 1);
     final int n2 = cutLength - n1;
-    final long b_end = corrected_ranks.correction[end - 1];
-    final long b_start = start == 0 ? 0 : corrected_ranks.correction[start - 1];
+    final long b_end = corrected_ranks.correction[safeEnd - 1];
+    final long b_start = safeStart == 0 ? 0 : corrected_ranks.correction[safeStart - 1];
     final double correction = (double) (b_end - b_start) / (cutLength * (cutLength - 1));
     final double std = FastMath.sqrt((((double) (n1 * n2)) / 12) * (cutLength + 1 - correction));
-    final double Z = std > 0 ? Math.abs((U - (0.5 * n1 * n2)) / std) : 0.;
+    if(std == 0) { return 0; }
```

Comment:
Note the std > 0 ternary in the previous version.

Reply:
Done